### DISCUSSION TITLE
Does Voicemail Express Amazon Connect support extension numbers similar to predicesor?

### DISCUSSION THREAD

**Discussion by Harkiiee** [amazon-connect/voicemail-express-amazon-connect#29] [Q&A]
Hey everyone!

I'm in the process of setting up an Amazon Connect instance for work. Previously, we used an older version that supported extension numbers, which are crucial to our workflow.

We used to use the one prior to these, however, the previous setup used Node.js 16, but I needed to upgrade to Node.js 20, and now everything has stopped working due to this, so I'm looking into this as an alternative.

Does anyone know if the latest version supports extension numbers? Or does anyone have suggestions for getting extension numbers working with this setup?

Thanks in advance!

**Comment by maekiteasy** [comment]
I'm not sure if this is exactly your situation, but it may give you some food for thought. In our setup, we have full DID numbers for each of our users, but refer to the last 4 digits as extensions. We point all of the numbers to a call flow that runs a Lambda script and chops the last 4 digits and determines the user or queue to transfer to.

I have also been practicing a migration path on our test environment going from the old Voicemail for Connect to the new VMX3. I have been playing with two methods: DynamoDB and Tags.

### Tags
- Within Amazon Connect, edit a user and scroll down. You will see the Tag section. You can add a key such as "Extension" and a value such as "2100" or whatever extensions you use. It doesn't have to be limited to 4 digits.
- This will simulate giving your users an extension. They wouldn't even technically need a full DID number backing them up, depending on how you do it.
- You will then have to create a Lambda function to search your Connect users until it finds the matching extension tag.

**The positive**: is this is very easy to manage. You can add the extensions within the Connect portal.
**The negative**: The Lambda function is slower in this method, since it has to search all users. This could take longer depending on the amount of users you have. (Increase your Lambda function timeout setting if needed).

### DynamoDB
- Create a new DynamoDB table that includes information such as primaryKey, agentLogin, agentFirstName, agentLastName, agentExtension. Or, keep the existing table that the old solution created (as long as it doesn't disappear when you delete the stack).
- Create a Lambda function to return agent info based on the extension passed to it.

**The positive**: if you build an index on the agentExtension, the lookups will be way faster than tags.
**The negative**: You will have to manually add entries for new users or extension changes to the DynamoDB table through the AWS console or use some other method.

I'll attach a sample Lambda function for Tags, but you'll have to let me know if any interest in DynamoDB. Or, have ChatGPT come up with one (it's pretty good at it). Once you figure out the agentLogin from the extension, set the working queue dynamically to that agent, along with the rest of the VMX3 variables, then invoke the VMX3 module.

```
// Lambda Tags function
// Pass a DID number from Connect flow to this Lambda function
// Requires agents to have Tag: Extension: ####
// Returns the last 4 digits as Extension as well as user details

import { ConnectClient, DescribeUserCommand, ListUsersCommand } from "@aws-sdk/client-connect";

export const handler = async (event) => {
    try {
        // Log the received event for debugging
        console.log("Incoming event:", JSON.stringify(event, null, 2));

        // Ensure event has ContactData
        if (!event.Details || !event.Details.ContactData || !event.Details.ContactData.Attributes) {
            throw new Error("Invalid event structure. 'ContactData' or 'Attributes' is missing.");
        }

        // Get the didNumber from the contact attributes
        const didNumber = event.Details.ContactData.Attributes.didNumber;
        if (!didNumber) {
            throw new Error("didNumber is missing from contact attributes.");
        }

        console.log("Received didNumber:", didNumber);

        // Extract the last 4 digits from the didNumber as the agentExtension
        const agentExtension = didNumber.slice(-4);
        console.log("Extracted agentExtension:", agentExtension);

        // Set up the Connect client
        const instanceId = "arn:aws:connect:instance";  // Replace with your actual Amazon Connect instance ID
        const client = new ConnectClient({});

        // List all users in the Amazon Connect instance
        const listUsersCommand = new ListUsersCommand({ InstanceId: instanceId });
        const users = await client.send(listUsersCommand);
        console.log("List of users fetched:", JSON.stringify(users, null, 2));  // Can comment out after tests if desired

        let agentDetails = null;

        // Iterate through the users and check their tags for the matching agentExtension
        for (const user of users.UserSummaryList) {
            const userArn = user.Arn;

            // Describe the user to get their tags and other details
            const describeUserCommand = new DescribeUserCommand({ InstanceId: instanceId, UserId: user.Id });
            const userDetails = await client.send(describeUserCommand);

            const userTags = userDetails.User.Tags;
            if (userTags && userTags.Extension === agentExtension) {
                const firstName = userDetails.User.IdentityInfo?.FirstName || "Unknown";
                const lastName = userDetails.User.IdentityInfo?.LastName || "Unknown";
                agentDetails = {
                    agentExtension: agentExtension,
                    agentLogin: userDetails.User.Username,
                    agentFullName: `${firstName} ${lastName}`,
                };
                break;
            }
        }

        if (agentDetails) {
            console.log("Agent details found:", JSON.stringify(agentDetails, null, 2));
            return agentDetails;
        } else {
            console.log("No agent found for the given extension:", agentExtension);
            throw new Error(`Agent with extension ${agentExtension} not found.`);
        }
    } catch (error) {
        console.error("Error occurred in Lambda function:", error);
        throw error;
    }
};
```

`
// Sample test event
{
  "Details": {
    "ContactData": {
      "Attributes": {
        "didNumber": "+18001234567"
      }
    }
  }
}`

**Comment by dippa69** [comment]
Will this ever be updated to support extension numbers. Its not feasible for large contact centers to have individual DID's. So many features are now gone with this new voicemail solution.

**Reply by dougjaso** [reply]
Voicemail Express is not an extenstion managements solution. If you need extensions or other per-user mapping, the current pattern would be to use the Data Tables feature in Amazon Connect, which does not require Lambda, Dynamo, or any other external integration.
